### PR TITLE
[JAX] Custom partitioning for Permutation primitives

### PR DESCRIPTION
**Pull request by tdophung**
# Description

Current JAX primitive for permutation does not have any partitioning or shardy rule. The approach here is to shard this along the B (batch axis) FSDP. 
Another sharding axis can be considered is EP, which might come in another PR, not this one. Will close the issue  #2536 once EP is implemented 

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:
- Add partitioning rules to permutation primitive
- Add test for distributed permutation op
- Modify the permute triton kernel in common to initialize all output data to 0s, to make sure padding values are all 0s, guaranteeing that multiplication with probs and grouped gemm that happens after will remain correct with padding

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/jax/test_distributed_permutation.py [hunk 1]
```diff
@@ -0,0 +1,597 @@
...
+            assert_allclose(jax.device_get(target_perm_probs), ref_perm_probs, dtype=dtype)
+
+            # Verify row_id_map n_routed column matches routing_map sum
+            target_rid_map_np = jax.device_get(target_rid_map)
+            assert jnp.array_equal(
+                target_rid_map_np[:, -1], ref_rid_map[:, -1]
+            ), "n_routed column mismatch"
+
+            # Sanity checks
+            target_out_np = jax.device_get(target_out)
```
Comment (jberchtold-nvidia) [inline tests/jax/test_distributed_permutation.py:260]: `jax.device_get` isn't required, calling any numpy function on a JAX array will implicitly call a JAX array's `__array__()` method and do this copying to host and conversion to a numpy array. But this doesn't negatively affect anything either so I'm okay with leaving this in for now so we can merge the PR

**Review by jberchtold-nvidia — APPROVED** [review]
LGTM, thanks!

File: tests/jax/test_distributed_permutation.py (hunk 1 above)
Comment (tdophung) [reply tests/jax/test_distributed_permutation.py:260]: Okay I will merge and address this in the next PR and keep this unresolved until so
